hydrolib/core/io/dflowfm/bc/models.py

```diff
@@ -617,12 +653,19 @@ def _validate_that_all_quantityunitpairs_have_valid_verticalpositionindex(
     def _validate_verticalpositionindexes_and_update_quantityunitpairs(
         verticalpositionindexes: List[int],
         number_of_verticalpositions: int,
-        quantityunitpairs: List[QuantityUnitPair],
+        quantityunitpairs: List[ScalarOrVectorQUP],
     ) -> None:
         if verticalpositionindexes is None:
             raise ValueError("vertPositionIndex is not provided")
 
-        if len(verticalpositionindexes) != len(quantityunitpairs) - 1:
+        all_qups = []
```

Comment:
I find it not so elegant that we need to create a full list, only to map a function across the individual list elements below.
I understand the cause: contrary to the original implementation, now *first* the VectorQUPs get packed by the base class, and only after that the verticalpositionindexes are added by the T3D class.

This is comment is for brainstorming only: I'm still thinking if there is a more elegant way or not.

Reply:
Discussed this with @tim-vd-aardweg in a brief meeting. We decided:
* this new `VectorForcingBase` class is still an improvement, so we keep it.
* the above duplicate list `all_qups` is either acceptable, or we change it as follows:
  * pass the original list `quantityunitpairs` to the method `_add_verticalpositionindex_to_quantityunitpairs()` and in there: don't do a `zip`, but instead loop across `quantityunitpairs` and when encountering a `VectorQuantityUnitPairs`, do a nested loop over its children, and for these do the actual assignment of `.verticalpositionindex=...`.
  * while looping+nested-looping, maintain a counter of total nr of regular qups, and only at the end do a check wether that is equal to the length of the input `verticalpositionindexes` list.

Reply:
I removed the creation of a separate list of qups. We now just iterate over all `ScalarOrVectorQUP` and unpack any `VectorQuantityUnitPairs` we find.